Fix #8285: napoleon can use PEP526 annotations

QuentinSoubeyran opened the pull request:
### Feature
Add an option so that napoleon fetches PEP 526-style  attributes annotation  in google style

### Purpose
This simplifies the use of the google style with PEP 526. The google style recommends adding the type of attributes in the class docstring only if the attribute doesn't have a type annotation in the code. As some libraries (e.g. `dataclasses`, `attrs`) uses those annotations and makes them mandatory, this features avoids duplicated type annotations.

### Detail
- added a new option to `napoleon`, `napoleon_google_attr_annotations` (boolean, defaults to False)
- If the option is `True`, in google style, if an attribute doesn't have a type in the docstring, the class is inspected for PEP 526-style annotations for that attribute.
- documented the new option in `napoleon.rst`
- added an example in the `example_google.py` in the documentation
- added a test case in `test/test_ext_napoleon_docstring.py`

*Note:* Default behaviour is not changed, as `napoleon_google_attr_annotations` defaults to False

### Relates
- #8285 

### Notes
I could only test python 3.8 and 3.9rc1, as `tox` couldn't find other versions.
Some `lualatex` test fails on my machine (including on the`3.x` branch so I don't think it's due to this change):
 - `tests/test_build_latex.py::test_build_latex_doc[lualatex-howto]`
- `tests/test_build_latex.py::test_build_latex_doc[lualatex-manual]`

I'm new to sphinx, so If I forgot anything I'm happy to improve this PR ! I'm not sure about the option name in particular, perhaps a better name can be found.

QuentinSoubeyran committed:
added napoleon_google_attr_annotations option to use PEP 526 on google style

QuentinSoubeyran committed:
corrected too long line for flake8

QuentinSoubeyran committed:
moved python 3.6+ syntax test data to standalone file

QuentinSoubeyran committed:
added behavior for exceptions

tk0miya commented:
Could you rebase this into 3.x branch? It seems you're trying to merge master branch (it is mainly for developing 4.x) to 3.x branch.

QuentinSoubeyran committed:
sorted imports

QuentinSoubeyran commented:
Hi, it should be good now, sorry for the mess. I used two force push to undo my erroneous merge and for the rebase, I hope this is fine.

There are still limitations with the code, I think: for instance, it does not comply with autodoc's `autodoc_type_aliases` config. I would expect type hints loaded from type annotations to be handled exactly like those autodoc derives from function signatures. They are very close in python code and have the same semantic.
Perhaps part of this feature should be implemented in autodoc ? The feature lies at the interface between autodoc and napoleon, as part code part docstring.

I'm open to suggestions or pointers on how to do that properly

QuentinSoubeyran commented:
When using my branch in my project (`pheres`), I'm getting weird `Handler <function warn_missing_reference at 0x...> for event 'warn-missing-reference' threw an exception (exception: 'NoneType' object has no attribute 'name')` errors, if there are missing references in my docstrings. It looks like the `domain` argument is None in [`domain/std.py`](https://github.com/QuentinSoubeyran/sphinx/blob/5e578d197865488696903b3e248a25ed1440d7a8/sphinx/domains/std.py#L1108).
I can't reproduce if I install sphinx using `pip install git+https://github.com/sphinx-doc/sphinx@3.x`.

I believe my `git rebase 3.x` went awry, because my branch reports running sphinx `v3.4.0+/fab8e7b`, while install from pip as above reports sphinx `v3.3.0+/fab8e7b`. I don't think I modified any code regarding missing reference or event handling.

I couldn't pinpoint if is a bug in `v3.4.0+/fab8e7b`, a bug in my code or a failed rebase. Any ideas ?

tk0miya commented on sphinx/ext/napoleon/docstring.py:
```diff
@@ -600,6 +601,27 @@ def _parse_attribute_docstring(self) -> List[str]:
     def _parse_attributes_section(self, section: str) -> List[str]:
         lines = []
         for _name, _type, _desc in self._consume_fields():
+            # code adapted from autodoc.AttributeDocumenter:add_directive_header
```
Now I posted https://github.com/sphinx-doc/sphinx/pull/8441 not to copy the function more.

tk0miya commented on sphinx/ext/napoleon/docstring.py:
```diff
@@ -600,6 +601,27 @@ def _parse_attribute_docstring(self) -> List[str]:
+                    # cache the class annotations
+                    if not hasattr(self, "_annotations"):
+                        try:
+                            self._annotations = get_type_hints(self._obj)
```
Please pass `self._config.autodoc_type_aliases` as the 3rd argument as same as I did in https://github.com/sphinx-doc/sphinx/pull/8435. Then it also supports `autodoc_type_aliases`!

I guess `napoleon_type_aliases` is also useful for this case.
https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html

QuentinSoubeyran replied:
I had to use `getattr()` as tests failed due to `AttributeError`s. Should I leave that there as a safety, or change the test to be properly configured (or both ?)

tk0miya commented on doc/usage/extensions/napoleon.rst:
```diff
@@ -511,3 +539,9 @@ sure that "sphinx.ext.napoleon" is enabled in `conf.py`::
 
    .. versionadded:: 3.2
+
+.. confval:: napoleon_google_attr_annotations
```
Now I'm thinking to apply this to numpy also. There is no reason to limit this feature to google style only.

QuentinSoubeyran replied:
I started looking at that, I'm not too sure how to do it yet. Is `sphinx.ext.napoleon.docstring.NumpyDocstring::_consume_field()` the right place ?

tk0miya commented on doc/usage/extensions/napoleon.rst:
```diff
@@ -511,3 +539,9 @@ sure that "sphinx.ext.napoleon" is enabled in `conf.py`::
+
+    True to allow using `PEP 526`_ attributes annotations in classes.
+    If an attribute is documented in the docstring without a type and
+    has an annotation in the class body, that type is used.
```
Please add `versionadded:: 3.4` after this block :-)

tk0miya commented on sphinx/ext/napoleon/__init__.py:
```diff
@@ -257,6 +258,9 @@ def __unicode__(self):
         section. If the entry is a tuple/list/indexed container, the first entry
         is the name of the section, the second is the section key to emulate.
 
+    napoleon_google_attr_annotations : :obj:`bool` (Defaults to False)
```
Is there any reason to disable this feature by default?

QuentinSoubeyran replied:
I was waiting on some sort of approval of the feature before enabling it by default. It is now enabled

tk0miya commented on sphinx/ext/napoleon/docstring.py:
```diff
@@ -600,6 +601,27 @@ def _parse_attribute_docstring(self) -> List[str]:
         lines = []
         for _name, _type, _desc in self._consume_fields():
+            # code adapted from autodoc.AttributeDocumenter:add_directive_header
+            if not _type and self._what in ('class', 'exception') and self._obj:
```
I think we can use this feature to module-level variables also.

QuentinSoubeyran replied:
What would be the value of `_what` for module-level variables ? Is it `module`, or another value not reflected in `GoogleDcostring` own docstring ?

QuentinSoubeyran replied:
Just understood what you meant, activated it for `module` too

tk0miya reviewed (changes requested):
Good work! I love this implementation. Let's make this better and merge into the next release!

tk0miya commented:
AFAIK, the errors around `warn-missing-reference` were added into the 3.x branch accidentally, but they have been resolved now. Please let me know if you still see the error.

QuentinSoubeyran committed:
refactored code and tests, prepared for numpy feature

QuentinSoubeyran commented:
Thank you for your review and help !

Following your comment on the `warn-missing-reference`, I think I rebased on the local `3.x` branch instead of the remote, without pulling first (oops). I rebased on the upstream after a fetch to fix that (this required another force push).

I also started implementing the requested changes:
- the feature is now called `napoleon_attr_annotations`, to prepare for numpy style implementation
- the feature is enabled by default
- refactored code to use `sphinx.util.typing.get_type_hints()`, with support for `config.napoleon_type_aliases` and `autodoc_type_aliases`
- added a small fix in `sphinx.util.typing.get_type_hints()` so that the `globalns` argument isn't ignored and forced to `None`. This is minor so I thought it would be OK to do in this PR instead of a new one.
- refactored tests to prepare testing module-level variables and numpy style implementation
- added the version in which the feature will be added in the doc

What is left to do is:
- implement the feature for numpy style (I haven't figured out exactly where to do that yet, see requested change discussion above)
- implement the feature for module-level variables (see requested change discussion above)

EDIT: added mention of the `get_type_hints` fix

QuentinSoubeyran committed:
implemented feature in numpy style, factored common code

QuentinSoubeyran committed:
enabled feature for module docstring

QuentinSoubeyran committed:
rebased on upstream/3.x; fixed CI

tk0miya commented on sphinx/ext/napoleon/__init__.py:
```diff
@@ -44,6 +44,7 @@ class Config:
         napoleon_preprocess_types = False
         napoleon_type_aliases = None
         napoleon_custom_sections = None
+        napoleon_attr_annotations = False
```
True by default?

QuentinSoubeyran replied:
Oops, thanks for catching that ! Fixed it

tk0miya commented on sphinx/ext/napoleon/docstring.py:
```diff
@@ -804,6 +808,21 @@ def _strip_empty(self, lines: List[str]) -> List[str]:
+            if self._what in ("module", "class", "exception") and self._obj:
+                # cache the class annotations
+                if not hasattr(self, "_annotations"):
+                    localns = getattr(self._config, "autodoc_type_aliases", {}) or {}
```
I feel this must be a bug of testing framework of napoleon. I feel this is ugly. But it's not related to this change. So let's fix in another PR.

QuentinSoubeyran replied:
Works for me. I believe this is because in napoleon test cases, the object passed as a`config` is an instance of napoleon's own `Config` class, which of course lacks autodoc's config options

QuentinSoubeyran replied:
Actually,  the `or {}` part is unnecessary. I assumed an unspecified option could be `None`, but the doc saying the default is `{}`. I removed `or {}` and tests and my own project using this feature work correctly.

tk0miya commented on sphinx/util/typing.py:
```diff
@@ -66,7 +66,7 @@ def get_type_hints(obj: Any, globalns: Dict = None, localns: Dict = None) -> Dic
 
     try:
-        return typing.get_type_hints(obj, None, localns)
+        return typing.get_type_hints(obj, globalns, localns)
```
Oops...

tk0miya reviewed (approved):
LGTM with nits.

QuentinSoubeyran committed:
fixed napoelon_attr_annotation documented as False by default

QuentinSoubeyran committed:
removed unnecessary 'or {}'

QuentinSoubeyran commented:
@tk0miya Fixed the documentation error and improved a bit the ugly code.
FYI, I still have the `warn-missing-reference` issue in my project with this branch, but also with the current `3.x` branch (commit 5a0e1231ad3fce9405ea52d90b21153e4ea86a27). If that helps, this is caused by the config line `default_role = "any"`. I'll do a proper bug report for that

EDIT: see #8491

tk0miya commented:
Merged. Thank you for your great work!

QuentinSoubeyran commented:
My pleasure !
